### DISCUSSION TITLE
next/link and component libraries

### DISCUSSION THREAD

**Discussion by deadlyhifi** [vercel/next.js#15560] [Help]
I'm currently working on a Design System and Component Library at my company. We're making re-usable React components (see [designsystemsrepo.com/design-systems](https://designsystemsrepo.com/design-systems/) for this type of thing).

Some of these components will take URLs as props. The question arises - how do we use next/link on an internal link? We need these components to not be tied to any implementation as they'll also be used in non next.js projects.

As I understand it next/link should only be used for routes managed by next.js, and [it fails if non managed URLs are sent in](https://github.com/vercel/next.js/blob/master/errors/invalid-href-passed.md). Therefore even if we came up with some kind of config to say these components are being used in next.js the link passed in would still need evaluating for internal or external (more difficult if not all routes on a domain are next.js managed).

Does anyone know of component libraries that handle this scenario, or are using a library and have figured out a way?

[nextjs.org/docs/api-reference/next/link](https://nextjs.org/docs/api-reference/next/link) could do with some clarification on when to use it. I couldn't see anything about not using it for external links.

And more of a discussion point, but would solve these problems - why require a next/link at all? At build time could all the routes managed by next.js be determined and all internal links have next/link put in automatically? With explicit use of next/link to supply config if it was a non default use-case?

**Comment by leerob** [comment]
Here's a link wrapper I use to handle both internal (next/link) and external links. I'm using [Chakra UI](https://chakra-ui.com/link)'s `Link`.

```js
const CustomLink = (props) => {
  const href = props.href;
  const isInternalLink = href && (href.startsWith('/') || href.startsWith('#'));

  if (isInternalLink) {
    return (
      <NextLink href={href} passHref>
        <Link {...props} />
      </NextLink>
    );
  }

  return <Link isExternal {...props} />;
};
```

> Why require a next/link at all? At build time could all the routes managed by next.js be determined and all internal links have next/link put in automatically? With explicit use of next/link to supply config if it was a non-default use-case?

If you have a _ton_ of links, you might not want to prefetch. It's probably better for the framework to allow you to choose.

**Reply by deadlyhifi** [reply]
Thanks for your response @leerob. To be clear then, you use this outside of the component library? Chakra, in this case, doesn't have any knowledge of it being used in nextjs, it merely renders the content you pass in.

My issue is more around an `href` being passed as a prop to a component.

```js
<SomeBanner href={href}>This entire component block will link, not just specific content within it</SomeBanner>
```

Would it be better as a component design decision to not accept a prop in that way -

```js
<SomeBanner><NextLink href={href}>{content}</NextLink></SomeBanner>
```
would be better?

Or even, as per [next/link](https://nextjs.org/docs/api-reference/next/link) docs:

```js
<NextLink href={href} passHref>
      <SomeBanner>{content}</SomeBanner>
</NextLink>
```

**Reply by mosherc** [reply]
I have run into this issue too. A quick and dirty way to do it might be to add a prop like `wrapper` that will wrap all links if it is defined. I'm guessing there is a more ideal way to do it.

**Reply by elwiiman** [reply]
Does somebody could figured out how to solve/implement this? 

Thanks!

**Comment by niek-ph** [comment]
bump

**Comment by sohamsshah** [comment]
Hi @deadlyhifi ,
I have found a workaround for this issue by the following approach.

In the UI-library, create a `Linker` Component as below:

In your `packages/ui/components/Linker/index.tsx`:
```
import React, { createContext } from "react";

export const LinkContext = createContext(null);
interface ILinker {
  CustomLink: any;

  children: React.ReactNode;
}
export const Linker = ({ CustomLink, children }: ILinker) => {
  return (
    <LinkContext.Provider value={CustomLink}>{children}</LinkContext.Provider>
  );
};
```
In the above approach, we create a context that stores the `CustomLink` component as the value of the LinkContext Provider. Meaning that it can be used at pretty much everywhere else in your component library by using `useContext`.

For example, you want to consume `LinkContext`'s value -> `CustomLink`.
You can do that as:
`packages/ui/components/Navbar/index.tsx`
```
export const Navbar = () => {
  const CustomLink: any = useContext(LinkContext);
  return (
      <nav>
             <ul>
                   <li> <CustomLink href="/">Home</CustomLink> </li>
                   <li> <CustomLink href="/discussions">Home</CustomLink> </li>
            </ul>
    </nav>
  );
};
```

How to use `Linker` in your Next App?
In `apps/web/pages/index.tsx`
```
import NextLink from "next/link"

const Index = (): JSX.Element => {
return (
      <Linker CustomLink={NextLink}> 
      // your ui library components here (that use next link internally)
      </Linker> 
      )
}
export default Index
```

In similar way, wrap your components with the `Linker` component to provide the NextLink across all the ui library.
You can do error handling for empty states when CustomLink value is undefined.

Cc: @leerob - Can you provide your inputs for this approach? It would be helpful to validate this idea and also improve it

**Reply by hugotox** [reply]
This looks cool, but the only issue I see is that `NextLink` uses custom props like `passHref` which is not a prop for react router's Link (if your component is planning to support both)

**Comment by tarunama** [comment]
bump

**Comment by der-daniel** [comment]
I was having the same issue... Here is what I came up with:

In my component library I have a component named `menu-item` which is just a styled `anchor`. On my library level I don't want to use `next/link` since my lib should be agnostic to the framework it gets embedded in.

The way I solved it, was defining a `Anchor` wrapper on my component which is to be implemented by the actual framework.

``` tsx
import React, { type FunctionComponent, type ReactNode } from 'react'

export type SidebarMenuItemProps = {
  text: string
  anchor?: ({ children }: { children: ReactNode }) => ReactNode
}

const SidebarMenuItem: FunctionComponent<SidebarMenuItemProps> = ({
  text,
  anchor,
}) => {
  const Anchor =
    anchor ?? (({ children }: { children: ReactNode }) => <>{children}</>)

  return (
    <>
      <Anchor>
        <div className="text-sm font-bold">{text}</div>
      </Anchor>
    </>
  )
}

export default SidebarMenuItem

```


In my `next.js` app I am using it like this.

``` tsx
<MenuItem
  anchor={({ children }: { children: ReactNode }) => (
    <Link href="/drafts">
      <a>{children}</a>
    </Link>
  )}
  text="This is a link"
/>
```

The little overhead of an inline component is worth it I think. This way I am also able to mix the `next/link` with anchor tags to overcome the tedeous trouble when I need a menu item to point to an external link.

Then I can use use a `anchor` tag.

``` tsx
<MenuItem
  anchor={({ children }: { children: ReactNode }) => (
    <a href="https://google.com" target="_blank">
      {children}
    </a>
  )}
  text="This is a link"
/>
```

I may adopt this pattern for all links. Forwarding the responsibility to the framework using the library is a good thing as many framework have their custom routing/linking/anchoring thing.

Hope this helps.

JMJ

**Comment by maxch-aa** [comment]
@leerob The `cloneElement` API, which next/link’s legacyBehaviour relies on, is a legacy React API method and should not be the main route to support anchor components in component libraries. See: https://github.com/reactjs/react.dev/issues/6057

next/link should export a hook that provides anchor element props that enable linking behaviour. Proposed API:

```ts
import { useLink } from 'next/link'
import CustomAnchor from 'component-system/anchor';

function MyLink({ label, href }) {
 const anchorProps = useLink({ href });
 return <CustomAnchor {...anchorProps}>{label}</CustomAnchor>;
}
```
